### PR TITLE
Quotient::Expected and QOlmExpected

### PR DESCRIPTION
**Pull request by KitsuneRal**
I have finally got to making a better wrapper than a generic `std::variant` for a combination of a value class and an error class. @TobiasFella, please let me know what you think. I'm going to proceed with merging your PR first so please don't rebase it, I'll rebase this one instead.

### CODE REVIEW

File: autotests/testolmsession.cpp [hunk 1]
```diff
@@ -45,22 +45,22 @@ void TestOlmSession::olmEncryptDecrypt()
     const auto encrypted = outboundSession->encrypt("Hello world!");
     if (encrypted.type() == QOlmMessage::PreKey) {
         QOlmMessage m(encrypted); // clone
-        QVERIFY(std::get<bool>(inboundSession->matchesInboundSession(m)));
+        QVERIFY(inboundSession->matchesInboundSession(m));
```
Comment (TobiasFella) [inline autotests/testolmsession.cpp:48]: IIUC this now just tests if the value is a bool, not if the bool is `true`

File: lib/expected.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#pragma once
```
Comment (TobiasFella) [inline lib/expected.h:4]: This is missing a copyright header

File: autotests/testolmutility.cpp [hunk 3]
```diff
@@ -79,7 +81,10 @@ void TestOlmUtility::verifySignedOneTimeKey()
...
 
     QOlmUtility utility2;
-    auto res2 = std::get<bool>(utility2.ed25519Verify(aliceOlm.identityKeys().ed25519, msg, signatureBuf1));
+    auto res2 =
+        utility2
+            .ed25519Verify(aliceOlm.identityKeys().ed25519, msg, signatureBuf1)
+            .value_or(false);
 
     //QCOMPARE(std::string(olm_utility_last_error(utility)), "SUCCESS");
     QCOMPARE(res2, true);
```
Comment (TobiasFella) [inline autotests/testolmutility.cpp]:
```suggestion
    QVERIFY(res2);
```

**Comment by TobiasFella** [comment]
+1, this is so much nicer than before

File: autotests/testolmsession.cpp (hunk 1 above)
Comment (KitsuneRal) [reply autotests/testolmsession.cpp:48]: No; I changed `matchesInboundSession()`. There should be no Olm error, and the value should be one.

**Comment by KitsuneRal** [comment]
...and I broke my own promise to first merge your PR :( sorry.

**Pull request merged**